components/Modals/SignUp.tsx

```diff
@@ -80,21 +99,53 @@ export default function SignUpModal({ open, onClose, onConfirm }) {
           {t('modal_signup.subheading_two')}
         </h2>
 
-        <pre className="mb-5 flex flex-col rounded-md border-2 border-dotted border-white/25 p-4">
-          <code className="mb-2 whitespace-pre-wrap break-all text-base">
-            {user ? user.privateKey.toString(16) : ''}
-          </code>
-          <CopyButton content={user ? user.privateKey.toString(16) : ''}>
-            {t('shared.copy')}
-          </CopyButton>
-        </pre>
+        <RadioGroup value={view} onChange={setView}>
+          <RadioButton name="generate" value={View.Generate}>
+            Generate
+          </RadioButton>
+
+          <RadioButton name="inpit" value={View.Input}>
+            Enter my own
+          </RadioButton>
+        </RadioGroup>
+
+        {view === View.Generate && (
+          <>
+            <pre className="mb-5 flex flex-col rounded-md border-2 border-dotted border-white/25 p-4">
+              {privateKey && (
+                <>
+                  <code className="mb-2 whitespace-pre-wrap break-all text-base">
+                    {privateKey.toString(16)}
+                  </code>
+                  <CopyButton content={privateKey.toString(16)}>
+                    {t('shared.copy')}
+                  </CopyButton>
+                </>
+              )}
+            </pre>
+            <p className="mt-5 text-base">{t('modal_signup.generate')}</p>
+          </>
+        )}
+
+        {view === View.Input && (
+          <>
+            <Input
+              type="text"
+              name="private_key"
+              placeholder="Enter your private key"
+              onInput={handleSetPrivateKey}
+            />
+            <p className="mt-5 text-base">{t('modal_signup.input')}</p>
+          </>
+        )}
 
-        <p className="mt-5 text-base">{t('modal_signup.paragraph_two')}</p>
         <button
-          onClick={confirm}
-          className="mt-4 w-full rounded-md border border-white px-4 py-2 text-xl text-white"
+          onClick={handleConfirm}
+          disabled={loading}
```

Comment:
Perhaps we should add a simple "I have backed up my private key checkbox" before allowing the user to proceed might help.

Also worth considering just immediately signing them in after this step. If they don't write down their private key they will still have to restart once they are logged out but worth considering

Reply:
I think the checkbox is a good idea, I'll add that. In general your second proposal is often considered a security risk but since we're not dealing with things like email verification I think we can do it.